daedaland/AthanasiaOtc.sol

```diff
@@ -0,0 +1,811 @@
+ *
+ * Typically during an NFT mint, configured value is deposited into this smart contract and an OTC purchase of the
+ * underlying token. This is done to avoid price spikes during minting.
+ * Alternatively, one may also choose to directly deposit the underlying tokens.
+ *
+ * This contract enables the OTC purchase of the underlying token. Only the IAthanasia contract may do the purchase, and only on
+ * behalf of the collection which has been registered with IAthanasiaOtc.
+ */
+interface IAthanasiaOtc {
+    struct TokenInfo {
+        address otcToken;
+        uint256 otcPrice;
+        uint256 totalAmount;
+    }
+
+    /**
+     * @dev Registers or updates a collection with the OTC contract.
+     *
+     * Parameters:
+     *  - `collection` - NFT Collection address on behalf of which the OTC purchase is done.
+     *  - `otcToken` - the token which is used to purchase the underlying token.
+     *  - `otcPrice` - the OTC price for 1 underlying token (i.e. 10eX, where X is number of decimals of underlying token)
+     *  - `totalAmount` - total amount of tokens that can be purchased on behalf of the collection.
+     */
+    function registerCollection(
+        address collection,
+        address otcToken,
+        uint256 otcPrice,
+        uint256 totalAmount
+    ) external;
+
+    /**
+     * @dev Returns true if the collection was registered with OTC contract for the specified OTC token and price.
+     *
+     * Parameters:
+     *  - `collection` - NFT Collection address on behalf of which the OTC purchase is done.
+     *  - `otcToken` - the token which is used to purchase the underlying token.
+     *  - `otcPrice` - the OTC price for 1 underlying token (i.e. 10eX, where X is number of decimals of underlying token)
+     *
+     * @return true of the registration data matches the parameters.
+     */
+    function validateCollection(
+        address collection,
+        address otcToken,
+        uint256 otcPrice
+    ) external returns (bool);
+
+    /**
+     * @dev Perform OTC purchase of `amount` amount of the underlying token, using the price and token type as registered for the collection.
+     *
+     * Parameters:
+     *  - `collection` - NFT Collection address on behalf of which the OTC purchase is done.
+     *  - `amountToPurchase` - amount of underlying tokens to purchase.
+     *  - `expectedCost` - total number of `otcToken` tokens that the caller expects to pay for the `amountToPurchase` underlying tokens.
+     */
+    function otc(
+        address collection,
+        uint256 amountToPurchase,
+        uint256 expectedCost
+    ) external payable;
+}
+
+interface IStakingHelper {
+    function stake(uint256 _amount, address _recipient) external;
+}
+
+interface IHectorTreasury {
+    function mintRewards(address _recipient, uint256 _amount) external;
+}
+
+contract AthanasiaOtc is IAthanasiaOtc, Ownable, Pausable {
```

Comment:
HectorAthanasiaOtc, there could be other protocols using it

Reply:
no, I meant the contract name. There could be more than one AthanasiaOtc

Reply:
> no, I meant the contract name. There could be more than one AthanasiaOtc

I see your point
Would you like to share your thought about it?